src/N3Store.js

```diff
@@ -762,6 +762,338 @@ export default class N3Store {
+    // return;
+    const changed = this._addToIndex(graphItem.subjects,   subject,   predicate, object);
+    if (!changed) return;
+    this._addToIndex(graphItem.predicates, predicate, object,    subject);
+    this._addToIndex(graphItem.objects,    object,    subject,   predicate);
+    // console.log(subject, predicate, object)
+    cb();
+  }
+
+  _addConclusion(conclusion, graph, cb) {
+    let c;
+    for (let i = 0; i < conclusion.length; i++) {
+      c = conclusion[i];
+      this._add(c.subject.value, c.predicate.value, c.object.value, graph, cb);
+    }
+  }
+
+  // TODO [FUTURE]: Improve performance by 'pre-computing' the index lookup
+  // e.g. if a rule only has one variable - then we can just give it a pointer
+  // to the index that it should do lookups from
+  // Similarly with insertions
+
+  _evaluatePremise(rule, content, cb, i = 0) {
+    let v0, v1, v2, value, [val0, val1, val2] = rule.premise[i].value, index = content[rule.premise[i].content], index1, index2;
+    v0 = !(value = val0.value);
+    for (value in v0 ? index : { [value]: index[value] }) {
+      if (index1 = index[value]) {
+        if (v0) val0.value = Number(value);
+        v1 = !(value = val1.value);
+        for (value in v1 ? index1 : { [value]: index1[value] }) {
+          if (index2 = index1[value]) {
+            if (v1) val1.value = Number(value);
+            v2 = !(value = val2.value);
+            for (value in v2 ? index2 : { [value]: index2[value] }) {
+              // TODO: Probably revert this
+              if (v2) val2.value = Number(value);
+
+              if (i === rule.premise.length - 1)
+                rule.conclusion.forEach(c => { 
+                  const changed = this._addToIndex(content.subjects,   c.subject.value,   c.predicate.value, c.object.value);
+                  if (!changed) return;
+                  this._addToIndex(content.predicates, c.predicate.value, c.object.value,    c.subject.value);
+                  this._addToIndex(content.objects,    c.object.value,    c.subject.value,   c.predicate.value);
+                  cb(c);
+                });
+              else
+                this._evaluatePremise(rule, content, cb, i + 1)
+            }
+            if (v2) val2.value = null;
+          }
+        }
+        if (v1) val1.value = null;
+      }
+    }
+    if (v0) val0.value = null;
+  }
+
+  _evaluateRules(rules, content, cb) {
+    for (let i = 0; i < rules.length; i++) {
+      this._evaluatePremise(rules[i], content, cb);
+    }
+  }
+
+  // A naive reasoning algorithm where rules are just applied by repeatedly applying rules
+  // until no more evaluations are made
+  _reasonGraphNaive(rules, content) {
+    // console.log('reasoning', rules)
+    
+    let add = true
+    // while (add) {
+    //   add = false
+    //   this._evaluateRules(rules, content, d => { 
+    //     add = true;
+    //   });
+    // }
+    
+    // return;
+    
+    let newRules = [];
```

Comment:
Thanks for checking @josd - this is where the memory error would be occurring. I can probably tweak the logic here so that rather than iterating over all the rules and then buffering newRules; I do this one rule at a time.


As noted earlier I can't actually download the 1_000_000 benchmark to test this - where do I find it?

Reply:
As a future note for myself, I can also reduce the amount that needs to be stored in the buffer by storing variable substitutions and a pointer to a rule, rather than all the generated conclusions.

Reply:
It is now at https://sourceforge.net/p/eulersharp/code/HEAD/tree/trunk/2009/12dtb/test-dl-1000000.n3
and also in the pull request https://github.com/jeswr/N3.js/pull/10

Reply:
Interestingly, it turns out that the problem is *not* in the length of this buffer, which in tests that I have run appears to not exceed a size of 3 on this benchmark.